src/Type/Symfony/SerializerInterfaceDynamicReturnTypeExtension.php

```diff
@@ -0,0 +1,45 @@
+<?php declare(strict_types = 1);
+
+namespace PHPStan\Type\Symfony;
+
+use PhpParser\Node\Expr\MethodCall;
+use PHPStan\Analyser\Scope;
+use PHPStan\Reflection\MethodReflection;
+use PHPStan\Type\ArrayType;
+use PHPStan\Type\Constant\ConstantStringType;
+use PHPStan\Type\DynamicMethodReturnTypeExtension;
+use PHPStan\Type\MixedType;
+use PHPStan\Type\ObjectType;
+use PHPStan\Type\Type;
+
+class SerializerInterfaceDynamicReturnTypeExtension implements DynamicMethodReturnTypeExtension
+{
+
+	public function getClass(): string
+	{
+		return 'Symfony\Component\Serializer\SerializerInterface';
+	}
+
+	public function isMethodSupported(MethodReflection $methodReflection): bool
+	{
+		return $methodReflection->getName() === 'deserialize';
+	}
+
+	public function getTypeFromMethodCall(MethodReflection $methodReflection, MethodCall $methodCall, Scope $scope): Type
+	{
+		$argType = $scope->getType($methodCall->args[1]->value);
+		if (!$argType instanceof ConstantStringType) {
+			return new MixedType();
+		}
+
+		$objectName = $argType->getValue();
+
+		if (substr($objectName, -2) === '[]') {
+			// The key type is determined by the data
+			return new ArrayType(new MixedType(false), new ObjectType(substr($objectName, 0, -2)));
```

Comment:
This part sould be recursive as you can define types like this `Foo[][]` and the denormalizer will handle them.

Reply:
You are right of course, https://github.com/phpstan/phpstan-symfony/pull/35